Doc/library/datetime.rst

```diff
@@ -1316,56 +1316,58 @@ Examples of working with datetime objects:
-    >>> class GMT1(tzinfo):
+    >>> from datetime import timedelta, datetime, tzinfo, timezone
+    >>> class KabulTz(tzinfo):
+    ...     # Kabul used +4 until 1945, when they moved to +4:30
+    ...     UTC_MOVE_DATE = datetime(1944, 12, 31, 20, tzinfo=timezone.utc)
     ...     def utcoffset(self, dt):
-    ...         return timedelta(hours=1) + self.dst(dt)
-    ...     def dst(self, dt):
-    ...         # DST starts last Sunday in March
-    ...         d = datetime(dt.year, 4, 1)   # ends last Sunday in October
-    ...         self.dston = d - timedelta(days=d.weekday() + 1)
-    ...         d = datetime(dt.year, 11, 1)
-    ...         self.dstoff = d - timedelta(days=d.weekday() + 1)
-    ...         if self.dston <=  dt.replace(tzinfo=None) < self.dstoff:
-    ...             return timedelta(hours=1)
+    ...         if dt.year < 1945:
+    ...             return timedelta(hours=4)
+    ...         elif (1945, 1, 1, 0, 0) <= dt.timetuple()[:5] < (1945, 1, 1, 0, 30):
+    ...             # If dt falls in the imaginary range, use fold to decide how
+    ...             # to resolve. See PEP495
+    ...             return timedelta(hours=4, minutes=(30 if dt.fold else 0))
     ...         else:
-    ...             return timedelta(0)
-    ...     def tzname(self,dt):
-    ...          return "GMT +1"
+    ...             return timedelta(hours=4, minutes=30)
     ...
-    >>> class GMT2(tzinfo):
-    ...     def utcoffset(self, dt):
-    ...         return timedelta(hours=2) + self.dst(dt)
-    ...     def dst(self, dt):
-    ...         d = datetime(dt.year, 4, 1)
-    ...         self.dston = d - timedelta(days=d.weekday() + 1)
-    ...         d = datetime(dt.year, 11, 1)
-    ...         self.dstoff = d - timedelta(days=d.weekday() + 1)
-    ...         if self.dston <=  dt.replace(tzinfo=None) < self.dstoff:
-    ...             return timedelta(hours=1)
+    ...     def fromutc(self, dt):
+    ...         # A custom implementation is required for fromutc as
+    ...         # the input to this function is a datetime with utc values
+    ...         # but with a tzinfo set to self
+    ...         # See datetime.astimezone or fromtimestamp
+    ...
+    ...         # Follow same validations as in datetime.tzinfo
+    ...         if not isinstance(dt, datetime):
+    ...             raise TypeError("fromutc() requires a datetime argument")
+    ...         if dt.tzinfo is not self:
+    ...             raise ValueError("dt.tzinfo is not self")
+    ...
+    ...         if dt.replace(tzinfo=timezone.utc) >= self.UTC_MOVE_DATE:
+    ...             return dt + timedelta(hours=4, minutes=30)
     ...         else:
-    ...             return timedelta(0)
+    ...             return dt + timedelta(hours=4)
+    ...
+    ...     def dst(self, dt):
+    ...         return timedelta(0)
+    ...
     ...     def tzname(self,dt):
-    ...         return "GMT +2"
+    ...         return "Asia/Kabul"
```

Comment:
I think this should return UTC+04:30 / UTC+04 depending on the time relative to the move date.

Reply:
from the docs:

> Return the time zone name corresponding to the datetime object dt, as a string. Nothing about string names is defined by the datetime module, and there’s no requirement that it mean anything in particular. For example, “GMT”, “UTC”, “-500”, “-5:00”, “EDT”, “US/Eastern”, “America/New York” are all valid replies

If the method is to return the timezone *name* I think it is probably better to have Kabul rather than the string representation of the offset.
Note the implementation of `timezone.utc` is "UTC"

Reply:
Traditionally, tzname() is returning something similar to strftime('%Z').  In the case of Kabul, this means

```
$ TZ=Asia/Kabul date -r 1000000000 +%Z
+0430
```
for recent times, and
```
$ TZ=Asia/Kabul date -r -1000000000 +%Z
+04
```
for times before the transition.  Paul Eggert, has been leading a [crusade on "invented" timezone abbreviations](https://mm.icann.org/pipermail/tz/2017-February/024837.html) recently resulting in %Z format becoming useless for many parts of the World.  I still believe tzname() should return a time-dependent abbreviation rather than a permanent zone name.

Reply:
OK! Updated via 56e7adf